### ISSUE TITLE
skip_serializing does not inhibit generic bounds

### ISSUE THREAD

**Issue by arcnmx** [serde-rs/serde#197]
The following will fail because the `derive` will automatically add a `T: Serialize/Deserialize` bound despite the attributes. It would be nice if the codegen were smart enough to figure out the bound isn't necessary.

Also note that `default` is not quite correct in this test case, as it doesn't imply that the type should never be deserialized. A `skip_deserializing` attribute should also exist (see also #183).

Also related: the `: Default` bound should not be necessary, and automatically added to the derived `impl Deserialize` in the same manner.

```rust
#![feature(plugin, custom_derive, custom_attribute)]
#![plugin(serde_macros)]

extern crate serde;

#[derive(Serialize, Deserialize)]
struct A<T: Default> {
	#[serde(skip_serializing, default)]
	t: T,
}

#[derive(Default)]
struct B;

fn main() {
	fn assert<T: serde::Serialize + serde::Deserialize>(_: T) { }

	assert(A { t: B })
}
```

**Comment by erickt** [comment]
@arcnmx: `#[serde(skip_deserializing)]` doesn't make sense since it'd be impossible to create the structure. It would be neat if we could skip adding bounds if we know need them.

**Comment by arcnmx** [comment]
@erickt the difference between `default` and `skip_deserializing` is that `default` still implies that type will be deserialized if possible. `skip_deserializing` would mean to *never even try*, just unconditionally initialize it from a given or default value and don't look for the key at all, etc.

**Comment by arcnmx** [comment]
(and would therefore mean a `T: Deserialize` bound is unneeded and wrong etc.)

**Comment by erickt** [comment]
Sure, that makes sense. Bumping this off the 0.7.0 release since it could be implemented backwards compatibly.

**Comment by dtolnay** [comment]
Using the changes in #285, the code above now compiles if you replace `#[serde(skip_serializing, default)]` with `#[serde(skip_serializing, skip_deserializing)]` (which is I think how @arcnmx intended for it to work).

**Comment by arcnmx** [comment]
@dtolnay awesome work, looks like that will indeed kill this issue off for good! I'd add that perhaps it's worth discussing whether `#[serde(skip_deserializing, default)]` should be required for consistency(ish) reasons, but I don't have a personal opinion on that anyway.

**Comment by dtolnay** [comment]
Thanks for bringing that up. My opinion is no - `skip_deserializing` should imply `default` unless you specify a different default with `default=...`.

**Comment by dtolnay** [comment]
This can be closed as all three PRs (#260, #265, #285) have merged. @arcnmx if you feel strongly about requiring an explicit `default` for `skip_deserializing`, please file a separate issue.

**Comment by erickt** [comment]
Thanks for the hard work @dtolnay!
